File: glyphtracy-ts/src/node.ts
import { clamp, dot, norm, unit } from "./math.js";
import { Contour } from "./contour.js";
import { AxisExtremaTag, Continuity, NodeTransition } from "./types.js";

export class Node {
  nodeId: number;
  contour: Contour;
  contourIndex: number;
  reasonTags: Set<string>;
  axisExtrema: Set<AxisExtremaTag>;
  transition: NodeTransition | null;
  continuity: Continuity | null;

  constructor(args: {
    nodeId: number;
    contour: Contour;
    contourIndex: number;
    reasonTags: Set<string>;
    axisExtrema?: Set<AxisExtremaTag>;
    transition?: NodeTransition | null;
    continuity?: Continuity | null;
  }) {
    this.nodeId = args.nodeId;
    this.contour = args.contour;
    this.contourIndex = args.contourIndex;
    this.reasonTags = args.reasonTags;
    this.axisExtrema = args.axisExtrema ?? new Set();
    this.transition = args.transition ?? null;
    this.continuity = args.continuity ?? null;
  }

  get pointRc(): number[] {
    return this.contour.pointsRc[this.contourIndex];
  }

  classifyContinuity(
    beforePointsRc: number[][],
    afterPointsRc: number[][],
  ): Continuity {
    if (this.transition !== "curve-curve") {
      return "non-continuous";
    }

    function weightedTangent(polyline: number[][]): number[] {
      if (polyline.length < 2) {
        return [1, 0];
      }
      const dirs: number[][] = [];
      for (let i = 0; i < polyline.length - 1; i += 1) {
        dirs.push(
          unit([
            polyline[i + 1][0] - polyline[i][0],
            polyline[i + 1][1] - polyline[i][1],
          ]),
        );
      }
      let vx = 0;
      let vy = 0;
      for (let i = 0; i < dirs.length; i += 1) {
        const w = 1 + i / Math.max(1, dirs.length - 1);
        vx += dirs[i][0] * w;
        vy += dirs[i][1] * w;
      }
      return unit([vx, vy]);
    }

    function localChain(polyline: number[][], fromEnd: boolean): number[][] {
      if (polyline.length <= 4) {
        return polyline;
      }

      const lengths: number[] = [];
      for (let i = 0; i < polyline.length - 1; i += 1) {
        lengths.push(
          norm([
            polyline[i + 1][0] - polyline[i][0],
            polyline[i + 1][1] - polyline[i][1],
          ]),
        );
      }
      const total = lengths.reduce((a, b) => a + b, 0);
      const target = Math.min(24, Math.max(8, 0.12 * total));

      if (fromEnd) {
        let acc = 0;
        let startIx = polyline.length - 2;
        while (startIx > 0 && acc < target) {
          acc += lengths[startIx];
          startIx -= 1;
        }
        return polyline.slice(startIx);
      }

      let acc = 0;
      let endIx = 1;
      while (endIx < polyline.length - 1 && acc < target) {
        acc += lengths[endIx - 1];
        endIx += 1;
      }
      return polyline.slice(0, endIx);
    }

    function signedCurvature(a: number[], b: number[], c: number[]): number {
      const ab = [b[0] - a[0], b[1] - a[1]];
      const bc = [c[0] - b[0], c[1] - b[1]];
      const ac = [c[0] - a[0], c[1] - a[1]];
      const abLen = norm(ab);
      const bcLen = norm(bc);
      const acLen = norm(ac);
      if (Math.min(abLen, bcLen, acLen) <= 1e-6) {
        return 0;
      }
      const cross = ab[0] * bc[1] - ab[1] * bc[0];
      return (2 * cross) / (abLen * bcLen * acLen);
    }

    if (beforePointsRc.length < 3 || afterPointsRc.length < 3) {
      return "non-continuous";
    }

    const beforeLocal = localChain(beforePointsRc, true);
    const afterLocal = localChain(afterPointsRc, false);
    if (beforeLocal.length < 3 || afterLocal.length < 3) {
      return "non-continuous";
    }

    const tanIn = weightedTangent(beforeLocal);
    const tanOut = weightedTangent(afterLocal);
    const turn = Math.acos(clamp(dot(tanIn, tanOut), -1, 1));
    if (turn > 0.35) {
      return "non-continuous";
    }

    const kIn = signedCurvature(
      beforeLocal[beforeLocal.length - 3],
      beforeLocal[beforeLocal.length - 2],
      beforeLocal[beforeLocal.length - 1],
    );
    const kOut = signedCurvature(afterLocal[0], afterLocal[1], afterLocal[2]);
    const kScale = Math.max(Math.abs(kIn), Math.abs(kOut), 1e-6);
    const kMatch = Math.abs(kIn - kOut) <= 0.06 + 0.35 * kScale;

    if (turn <= 0.14 && kMatch) {
      return "G2";
    }
    return "G1";
  }

  debugEntry(): Record<string, unknown> {
    return {
      node_id: this.nodeId,
      contour_id: this.contour.contourId,
      contour_index: this.contourIndex,
      point_rc: [this.pointRc[0], this.pointRc[1]],
      reasons: [...this.reasonTags].sort(),
      axis_extrema: [...this.axisExtrema].sort(),
      transition: this.transition,
      continuity: this.continuity,
    };
  }
}

File: Lib/glyphtracy/node.py
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Literal, Optional

if TYPE_CHECKING:
    from glyphtracy.contour import Contour

import numpy as np

NodeTransition = Literal["curve-curve", "curve-line", "line-line"]
Continuity = Literal["G1", "G2", "non-continuous"]
SegmentKind = Literal["line", "bezier"]
AxisExtremaTag = Literal["x", "y"]


@dataclass
class Node:
    node_id: int
    contour: "Contour"
    contour_index: int
    reason_tags: set[str]
    axis_extrema: set[AxisExtremaTag] = field(default_factory=set)
    transition: Optional[NodeTransition] = None
    continuity: Optional[Continuity] = None

    @property
    def point_rc(self) -> np.ndarray:
        return self.contour.points_rc[self.contour_index]

    def classify_continuity(
        self,
        before_points_rc: np.ndarray,
        after_points_rc: np.ndarray,
    ) -> Continuity:
        if self.transition != "curve-curve":
            return "non-continuous"

        def _normalized(vec: np.ndarray) -> np.ndarray:
            norm = float(np.linalg.norm(vec))
            if norm <= 1e-9:
                return np.array([1.0, 0.0], dtype=np.float64)
            return vec / norm

        def _weighted_tangent(polyline: np.ndarray) -> np.ndarray:
            if polyline.shape[0] < 2:
                return np.array([1.0, 0.0], dtype=np.float64)
            segments = polyline[1:] - polyline[:-1]
            dirs = np.array([_normalized(seg) for seg in segments], dtype=np.float64)
            weights = np.linspace(1.0, 2.0, dirs.shape[0], dtype=np.float64)
            vec = np.sum(dirs * weights[:, None], axis=0)
            return _normalized(vec)

        def _local_chain(polyline: np.ndarray, *, from_end: bool) -> np.ndarray:
            if polyline.shape[0] <= 4:
                return polyline

            segment_lengths = np.linalg.norm(polyline[1:] - polyline[:-1], axis=1)
            total_length = float(np.sum(segment_lengths))
            target_length = min(24.0, max(8.0, 0.12 * total_length))

            if from_end:
                acc_length = 0.0
                start_ix = polyline.shape[0] - 2
                while start_ix > 0 and acc_length < target_length:
                    acc_length += float(segment_lengths[start_ix])
                    start_ix -= 1
                return polyline[start_ix:]

            acc_length = 0.0
            end_ix = 1
            while end_ix < polyline.shape[0] - 1 and acc_length < target_length:
                acc_length += float(segment_lengths[end_ix - 1])
                end_ix += 1
            return polyline[:end_ix]

        def _signed_curvature(a: np.ndarray, b: np.ndarray, c: np.ndarray) -> float:
            ab = b - a
            bc = c - b
            ac = c - a
            ab_len = float(np.linalg.norm(ab))
            bc_len = float(np.linalg.norm(bc))
            ac_len = float(np.linalg.norm(ac))
            if min(ab_len, bc_len, ac_len) <= 1e-6:
                return 0.0
            cross = float(ab[0] * bc[1] - ab[1] * bc[0])
            return 2.0 * cross / (ab_len * bc_len * ac_len)

        if before_points_rc.shape[0] < 3 or after_points_rc.shape[0] < 3:
            return "non-continuous"

        before_local = _local_chain(before_points_rc, from_end=True)
        after_local = _local_chain(after_points_rc, from_end=False)
        if before_local.shape[0] < 3 or after_local.shape[0] < 3:
            return "non-continuous"

        tan_in = _weighted_tangent(before_local)
        tan_out = _weighted_tangent(after_local)

        dot = float(np.dot(tan_in, tan_out))
        dot = max(-1.0, min(1.0, dot))
        turn = float(np.arccos(dot))
        if turn > 0.35:
            return "non-continuous"

        k_in = _signed_curvature(
            before_local[-3],
            before_local[-2],
            before_local[-1],
        )
        k_out = _signed_curvature(
            after_local[0],
            after_local[1],
            after_local[2],
        )
        k_scale = max(abs(k_in), abs(k_out), 1e-6)
        k_match = abs(k_in - k_out) <= (0.06 + 0.35 * k_scale)

        if turn <= 0.14 and k_match:
            return "G2"
        return "G1"

    def debug_entry(self) -> dict:
        point_rc = self.point_rc
        return {
            "node_id": int(self.node_id),
            "contour_id": int(self.contour.contour_id),
            "contour_index": int(self.contour_index),
            "point_rc": [float(point_rc[0]), float(point_rc[1])],
            "reasons": sorted(self.reason_tags),
            "axis_extrema": sorted(self.axis_extrema),
            "transition": self.transition,
            "continuity": self.continuity,
        }
